### PR TITLE
Added error handling for MEL.ILogger related to context data

### PR DESCRIPTION
**Pull request by tippmar-nr**
Thank you for submitting a pull request.  Please review our [contributing guidelines](/CONTRIBUTING.md) and [code of conduct](https://opensource.newrelic.com/code-of-conduct/).

## Description

* Added error handling in `MicrosoftLoggingWrapper.GetContextData()` for cases where a logger implements `MEL.ILogger` but doesn't also expose an `IExternalScopeProvider` property named `ScopeProvider`. If that happens, we log a message and disable further attempts at collecting context data.
* Added integration tests and a dummy implementation of `MEL.ILogger` to verify the fix is working as expected.

# Author Checklist
- [x] Unit tests, Integration tests, and Unbounded tests completed
- [x] Performance testing completed with satisfactory results (if required)
- [x] [Agent Changelog](/src/Agent/CHANGELOG.md) or [Lambda Agent changelog](/src/AwsLambda/CHANGELOG.md) updated 

# Reviewer Checklist
- [ ] Perform code review
- [ ] Pull request was adequately tested (new/existing tests, performance tests)
- [ ] Review Changelog

### CODE REVIEW

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/MicrosoftExtensionsLogging/MicrosoftLoggingWrapper.cs [hunk 1]
```diff
@@ -53,52 +55,64 @@ private void RecordLogMessage(MethodCall methodCall, MEL.ILogger logger, IAgent
...
             }
         }
 
-        private static Dictionary<string, object> GetContextData(MEL.ILogger logger)
+        private static Dictionary<string, object> GetContextData(MEL.ILogger logger, IAgent agent)
         {
-            // We are trying to access this property:
-            // logger.Loggers[0].Logger.ScopeProvider
+            if (_contextDataNotSupported) // short circuit if we previously got an exception trying to access context data
+                return null;
```
Comment (JcolemanNR) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/MicrosoftExtensionsLogging/MicrosoftLoggingWrapper.cs:68]: Pedantic feedback.. our coding standard is supposed to include `{ }` even for one line control logic

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/MicrosoftExtensionsLogging/MicrosoftLoggingWrapper.cs [hunk 2]
```diff
@@ -53,52 +55,64 @@ private void RecordLogMessage(MethodCall methodCall, MEL.ILogger logger, IAgent
...
-            }, harvestedKvps);
+                    // Possibly handle case of IEnumerable<KeyValuePair<object, object>>, etc (not now though)
+                }, harvestedKvps);
 
-            return harvestedKvps;
+                return harvestedKvps;
+            }
+            catch (Exception e)
+            {
+                agent.Logger.Log(Level.Debug, $"Unexpected exception while attempting to get context data. Context data is not supported for this logging framework. Exception: {e}");
```
Comment (JcolemanNR) [inline src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/MicrosoftExtensionsLogging/MicrosoftLoggingWrapper.cs:112]: I would log this at warning so customers don't need to turn logs up when they wonder why their context isn't working

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+            _fixture.AddActions
+            (
+                setupConfiguration: () =>
+                {
+                    var configModifier = new NewRelicConfigModifier(fixture.DestinationNewRelicConfigFilePath);
+
+                    configModifier
+                    .EnableLogMetrics(true)
+                    .EnableLogForwarding(true)
```
Comment (JcolemanNR) [inline tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:61]: Already enabled by default

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    .EnableContextData(_contextDataEnabled)
+                    .EnableDistributedTrace()
```
Comment (JcolemanNR) [inline tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:63]: already enabled by default

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            );
+
+            _fixture.Initialize();
+        }
+
+        [Fact]
+        public void Test()
+        {
+            var match = _fixture.AgentLog.TryGetLogLines(AgentLogBase.ContextDataNotSupportedLogLineRegex);
+            Assert.Single(match);
```
Comment (JcolemanNR) [inline tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:79]: Test needs to assert that logs were actually sent as well. Checking that we logged about context data being disabled is different that verifying we still forward logs.

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+            { "mycontext3", "test" },
+            { "mycontext4", "value" },
+        };
+
+
+        public ContextDataNotSupportedTestsBase(TFixture fixture, ITestOutputHelper output, bool contextDataEnabled, LoggingFramework loggingFramework) : base(fixture)
+        {
+            _fixture = fixture;
+            _loggingFramework = loggingFramework;
+            _contextDataEnabled = contextDataEnabled;
```
Comment (JcolemanNR) [inline tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:39]: All the test cases pass false to `contextDataEnabled` in the ctor. We should add test coverage for the new logger impl here, or in `ContextDataTests.cs`

**Review by JcolemanNR — CHANGES_REQUESTED** [review]
Looking good! I present you with some requested changes, good sir! :)

File: src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/MicrosoftExtensionsLogging/MicrosoftLoggingWrapper.cs (hunk 1 above)
Comment (tippmar-nr) [reply src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/MicrosoftExtensionsLogging/MicrosoftLoggingWrapper.cs:68]: Thanks - first time I'm hearing there's a coding standard. Looked around and found it in the team wiki (where I probably should have noticed it before), but didn't see anything specifically addressing the need for curly braces in this case. I've added a bullet point in the WB Design Meeting for this week to discuss updating the coding standards.

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs (hunk 5 above)
Comment (tippmar-nr) [reply tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:79]: Good point. Updated to verify logs were forwarded but without any context data.

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs (hunk 6 above)
Comment (tippmar-nr) [reply tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:39]: I removed the `contextDataEnabled` parameter as it's not necessary for this specific test case; context data will be explicitly enabled for the tests

File: tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs (hunk 3 above)
Comment (tippmar-nr) [reply tests/Agent/IntegrationTests/IntegrationTests/Logging/ContextDataNotSupportedTests.cs:61]: Copy / pasted from the existing `ContextDataTests.cs` file, assumed those lines were written with purpose.  Will remove the unnecessary lines.

**Review by chynesNR — APPROVED** [review]
Looks good!

**Review by JcolemanNR — APPROVED** [review]

**Pull request merged**